src/mesh/Utils.hpp

```diff
@@ -132,5 +153,67 @@ std::array<Eigen::VectorXd, n> coordsFor(const std::array<Vertex *, n> &vertexPt
   return coords;
 }
 
+/// Given the data and the mesh, this function returns the surface integral. Assumes no overlap exists for the mesh
+inline Eigen::VectorXd integrate(PtrMesh mesh, PtrData data)
+{
+  const int       valueDimensions = data->getDimensions();
+  const int       meshDimensions  = mesh->getDimensions();
+  const auto &    values          = data->values();
+  Eigen::VectorXd integral        = Eigen::VectorXd::Zero(valueDimensions);
+
+  if (meshDimensions == 2) {
+    for (const auto &edge : mesh->edges()) {
+      int vertex1 = edge.vertex(0).getID() * valueDimensions;
+      int vertex2 = edge.vertex(1).getID() * valueDimensions;
+      for (int dim = 0; dim < valueDimensions; ++dim) {
+        integral(dim) += 0.5 * edgeLength(edge) * (values(vertex1 + dim) + values(vertex2 + dim));
+      }
+    }
+  } else {
+    for (const auto &face : mesh->triangles()) {
+      int vertex1 = face.vertex(0).getID() * valueDimensions;
+      int vertex2 = face.vertex(1).getID() * valueDimensions;
+      int vertex3 = face.vertex(2).getID() * valueDimensions;
+      for (int dim = 0; dim < valueDimensions; ++dim) {
+        integral(dim) += (triangleArea(face) / 3.0) * (values(vertex1 + dim) + values(vertex2 + dim) + values(vertex3 + dim));
+      }
+    }
+  }
+  return std::move(integral);
+}
+
+/// Given the data and the mesh, this function returns the surface integral
+inline Eigen::VectorXd integrateOverlap(PtrMesh mesh, PtrData data)
+{
+  const int       valueDimensions = data->getDimensions();
+  const int       meshDimensions  = mesh->getDimensions();
+  const auto &    values          = data->values();
+  Eigen::VectorXd integral        = Eigen::VectorXd::Zero(valueDimensions);
+
+  if (meshDimensions == 2) {
+    for (const auto &edge : mesh->edges()) {
+      int vertex1 = edge.vertex(0).getID() * valueDimensions;
+      int vertex2 = edge.vertex(1).getID() * valueDimensions;
+      if (edge.vertex(0).isOwner() and edge.vertex(1).isOwner()) {
+        for (int dim = 0; dim < valueDimensions; ++dim) {
+          integral(dim) += 0.5 * edgeLength(edge) * (values(vertex1 + dim) + values(vertex2 + dim));
+        }
+      }
+    }
+  } else {
+    for (const auto &face : mesh->triangles()) {
+      int vertex1 = face.vertex(0).getID() * valueDimensions;
+      int vertex2 = face.vertex(1).getID() * valueDimensions;
+      int vertex3 = face.vertex(2).getID() * valueDimensions;
+      if (face.vertex(0).isOwner() and face.vertex(1).isOwner() and face.vertex(2).isOwner()) {
+        for (int dim = 0; dim < valueDimensions; ++dim) {
+          integral(dim) += (triangleArea(face) / 3.0) * (values(vertex1 + dim) + values(vertex2 + dim) + values(vertex3 + dim));
+        }
+      }
+    }
+  }
+  return std::move(integral);
```

Comment:
```suggestion
  return integral;
```

This is a common antipattern which works against the optimizer.